### ISSUE TITLE
Can I pivot to more conventional Go template relationships?

### ISSUE THREAD

**Issue by brandon-rhodes** [livegrep/livegrep#170]
A very quick question, if you have a moment:

My recent CSP work eliminated errors for the front page (yay!), but not the file view page — which makes perfect sense, when I went to look: I had overlooked the fact that `fileview.html` has several `<script>` tags of its own, separate from those in `layout.html`.

So I need to inject a CSP script nonce into both levels: the layout template *and* the file view template.

The awkwardness is that the two templates are rendered in completely separate code paths. The file view is rendered inline in the server code with `executeTemplate(s.T.FileView, data)` whereas the layout that wraps it is then rendered with the custom `renderPage()` function. To provide nonces at both levels, I'd currently have to go edit both of those call sites, passing around extra data to make sure they can both access the nonce string.

I can certainly make the modification in both places. But:

* I became tempted to try sharing the code between the two sites somehow.
* Which made me wonder why the templates are rendered in separate steps at all. Go templates can call each other — why doesn't the layout template have `{{template...}}` constructs that pages like the search and file view provide their own ``{{define...}}`` blocks for?
* Which made me wonder why templates are loaded by reflection, instead of with Go's native ``t.ParseGlob`` mechanism that would then let the templates refer to each other by name.

I daren't plunge into any refactorings without understanding the history of how the current template code got to its current state, of course — could you shed any light on why template rendering happens in several stages, each of which will need the nonce, instead of in a single go with normal template inheritance?

Thanks for any advice you can provide!

**Comment by brandon-rhodes** [comment]
(And: if you're open to the template rendering being simplified and unified, I'll be happy to submit it as a series of small obvious pull requests; I needn't drop a single huge changes-everything-and-adds-CSP-nonces-at-the-same-time pull request on the project!)

**Comment by nelhage** [comment]
I'm very open to a refactor! This was one of my first projects using Go templating and I made it upas I went along.

I don't recall `ParseGlob`; Is it possible it was new since I wrote this code?

One problem I recall is that I think I tried one naive approach where the layout template had `{{template}}`` and the content pages provided `{{define}}`s, but then you couldn't load them all into the same template object, because of the conflicting `{{define}}`s with the same name, so you ended  up needing to build a separate template object for each view, and loading the layout template separately for each.

But I have no attachment to the status quo and I'd happily review PRs to modernize it!
